File: disk_agent/diff.py
from __future__ import annotations

from dataclasses import dataclass
from pathlib import Path
from typing import Dict, List, Tuple

from .models import DirectoryUsage, Snapshot, load_snapshot
from .report import format_bytes, snapshot_paths
from .snapshot import SNAPSHOT_DIR

SIGNIFICANT_BYTES = 50 * 1024 * 1024


@dataclass
class UsageChange:
    path: str
    bytes: int


def latest_two(directory: Path = SNAPSHOT_DIR) -> Tuple[Snapshot, Snapshot]:
    paths = snapshot_paths(directory)
    if len(paths) < 2:
        raise RuntimeError("two snapshots are required; snapshots are stored once per day")
    return load_snapshot(paths[-2]), load_snapshot(paths[-1])


def _usage_map(snapshot: Snapshot) -> Dict[str, int]:
    result: Dict[str, int] = {}
    for items in (snapshot.home_usage, snapshot.local_share_usage, snapshot.copilot_usage):
        for item in items:
            result[item.path] = item.bytes
    return result


def compare_snapshots(before: Snapshot, after: Snapshot) -> List[UsageChange]:
    old = _usage_map(before)
    new = _usage_map(after)
    changes = [
        UsageChange(path, new.get(path, 0) - old.get(path, 0))
        for path in old.keys() | new.keys()
    ]
    return sorted(
        (change for change in changes if abs(change.bytes) >= SIGNIFICANT_BYTES),
        key=lambda change: abs(change.bytes),
        reverse=True,
    )


def render_diff(before: Snapshot, after: Snapshot) -> str:
    changes = compare_snapshots(before, after)
    growth = [change for change in changes if change.bytes > 0]
    shrinkage = [change for change in changes if change.bytes < 0]
    lines = ["Compared:", f"{before.timestamp[:10]} → {after.timestamp[:10]}", "", "Growth:", ""]
    lines.extend(f"{format_bytes(change.bytes, signed=True)} {change.path}" for change in growth)
    if not growth:
        lines.append("No significant growth.")
    lines.extend(["", "Shrinkage:", ""])
    if shrinkage:
        lines.extend(f"{format_bytes(change.bytes, signed=True)} {change.path}" for change in shrinkage)
    else:
        lines.append("No significant shrinkage.")
    lines.extend(["", "No other significant changes."])
    return "\n".join(lines)


def diff_command() -> str:
    return render_diff(*latest_two())

File: src/diff.rs
use std::collections::{BTreeSet, HashMap};
use std::path::Path;

use anyhow::{anyhow, Result};

use crate::json::load_snapshot;
use crate::models::{Snapshot, UsageChange};
use crate::output::format_bytes;
use crate::paths;
use crate::report::snapshot_paths;

pub const SIGNIFICANT_BYTES: i64 = 50 * 1024 * 1024;

pub fn latest_two_from(directory: &Path) -> Result<(Snapshot, Snapshot)> {
    let paths = snapshot_paths(directory)?;
    if paths.len() < 2 {
        return Err(anyhow!(
            "two snapshots are required; snapshots are stored once per day"
        ));
    }
    let before = load_snapshot(&paths[paths.len() - 2])?;
    let after = load_snapshot(&paths[paths.len() - 1])?;
    Ok((before, after))
}

pub fn compare_snapshots(before: &Snapshot, after: &Snapshot) -> Vec<UsageChange> {
    let old = usage_map(before);
    let new = usage_map(after);
    let paths = old
        .keys()
        .chain(new.keys())
        .cloned()
        .collect::<BTreeSet<_>>();
    let mut changes = paths
        .into_iter()
        .map(|path| UsageChange {
            bytes: new.get(&path).copied().unwrap_or(0) - old.get(&path).copied().unwrap_or(0),
            path,
        })
        .filter(|change| change.bytes.abs() >= SIGNIFICANT_BYTES)
        .collect::<Vec<_>>();
    changes.sort_by(|left, right| right.bytes.abs().cmp(&left.bytes.abs()));
    changes
}

pub fn render_diff(before: &Snapshot, after: &Snapshot) -> String {
    let changes = compare_snapshots(before, after);
    let growth = changes
        .iter()
        .filter(|change| change.bytes > 0)
        .collect::<Vec<_>>();
    let shrinkage = changes
        .iter()
        .filter(|change| change.bytes < 0)
        .collect::<Vec<_>>();

    let mut lines = vec![
        "Compared:".to_string(),
        format!("{} → {}", &before.timestamp[..10], &after.timestamp[..10]),
        String::new(),
        "Growth:".to_string(),
        String::new(),
    ];
    if growth.is_empty() {
        lines.push("No significant growth.".to_string());
    } else {
        lines.extend(
            growth.iter().map(|change| {
                format!("{} {}", format_bytes(Some(change.bytes), true), change.path)
            }),
        );
    }

    lines.extend([String::new(), "Shrinkage:".to_string(), String::new()]);
    if shrinkage.is_empty() {
        lines.push("No significant shrinkage.".to_string());
    } else {
        lines.extend(
            shrinkage.iter().map(|change| {
                format!("{} {}", format_bytes(Some(change.bytes), true), change.path)
            }),
        );
    }
    lines.extend([String::new(), "No other significant changes.".to_string()]);
    lines.join("\n")
}

pub fn diff_command() -> Result<String> {
    let directory = paths::snapshot_dir()?;
    let (before, after) = latest_two_from(&directory)?;
    Ok(render_diff(&before, &after))
}

fn usage_map(snapshot: &Snapshot) -> HashMap<String, i64> {
    let mut result = HashMap::new();
    for items in [
        &snapshot.home_usage,
        &snapshot.local_share_usage,
        &snapshot.copilot_usage,
    ] {
        for item in items {
            result.insert(item.path.clone(), item.bytes);
        }
    }
    result
}
